File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/clustergroup/ClusterGroup.scala
package net.gree.aurora.scala.domain.clustergroup

import net.gree.aurora.domain.clustergroup.ClusterGroupFactory
import net.gree.aurora.domain.clustergroup.{ClusterGroup => JClusterGroup}
import net.gree.aurora.scala.domain.cluster.{Cluster, ClusterId}
import org.sisioh.dddbase.core.model.Entity
import scala.collection.JavaConverters._
import scala.language.implicitConversions
import scala.util.Try

/**
 * [[net.gree.aurora.domain.clustergroup.ClusterGroup]]のScala版。
 */
trait ClusterGroup extends Entity[ClusterGroupId] {

  /**
   *
   * @param shardId
   * @return
   */
  def resolveCluster(shardId: ClusterId): Try[Cluster]

  def clusters: Try[Seq[Cluster]]

  override def toString =
    Seq(s"clusters = $clusters").mkString("ClusterGroup(", ", ", ")")

}

/**
 * コンパニオンオブジェクト。
 */
object ClusterGroup {

  private[scala] def apply(underlying: JClusterGroup): ClusterGroup =
    new ClusterGroupImpl(underlying)

  /**
   * ファクトリメソッド。
   *
   * @param identity 識別子
   * @param clusters [[net.gree.aurora.scala.domain.cluster.Cluster]]の集合　
   * @return [[net.gree.aurora.scala.domain.clustergroup.ClusterGroup]]
   */
  def apply(identity: ClusterGroupId, clusters: Seq[Cluster]): ClusterGroup =
    apply(ClusterGroupFactory.create(identity, clusters.map(Cluster.toJava).asJava))

  /**
   * ScalaオブジェクトからJavaオブジェクトに変換する。
   *
   * @param self [[net.gree.aurora.scala.domain.clustergroup.ClusterGroup]]
   * @return [[net.gree.aurora.domain.clustergroup.ClusterGroup]]
   */
  implicit def toJava(self: ClusterGroup) = self match {
    case s: ClusterGroupImpl => s.underlying
  }

}

File: aurora-core/src/main/java/net/gree/aurora/domain/clustergroup/ClusterGroup.java
package net.gree.aurora.domain.clustergroup;

import net.gree.aurora.domain.cluster.Cluster;
import net.gree.aurora.domain.cluster.ClusterId;
import org.sisioh.dddbase.model.Entity;
import org.sisioh.dddbase.utils.Try;

import java.util.List;

/**
 * {@link Cluster}の集合を表すエンティティ。
 */
public interface ClusterGroup
        extends Entity<ClusterGroupId>, Comparable<ClusterGroup> {

    /**
     * {@link ClusterId}から{@link Cluster}を解決する。
     *
     * @param clusterId {@link ClusterId}
     * @return Tryでラップした{@link Cluster}
     */
    Try<Cluster> resolveCluster(ClusterId clusterId);

    /**
     * {@link Cluster}の集合を取得する。
     *
     * @return Tryでラップされた{@link Cluster}のリスト
     */
    Try<List<Cluster>> getClusters();

    /**
     * オーダーを取得する。
     *
     * @return オーダー
     */
    Long getOrder();

    /**
     * インスタンスを複製する。
     *
     * @return 複製したインスタンス
     */
    ClusterGroup clone();

}
